File: src/State/MainMenu.cpp
#include "MainMenu.hpp"

#include "../Console/Console.hpp"
#include "../GUI/Menu.hpp"
#include "../GUI/Window.hpp"
#include "../GUIMenu/OptionsMenu.hpp"
#include "../Input/Event.hpp"
#include "../Input/Input.hpp"
#include "../Lua/Lua.hpp"
#include "../Resource/ResourceManager.hpp"
#include "../Utils/Asset.hpp"
#include "../Utils/CFG.hpp"
#include "../Utils/Utils.hpp"

using mmm::vec2;

MainMenu::MainMenu(Asset* asset) {
  setLoggerName("MainMenu");
  mAsset = asset;
  mLua   = mAsset->lua();
  mAsset->rManager()->unloadUnnecessary(ResourceScope::MainMenu);
  mAsset->rManager()->loadRequired(ResourceScope::MainMenu);

  OptionsMenu* opts = new OptionsMenu(mAsset->input());
  CFG*         cfg  = mAsset->cfg();
  Window*      menu =
    new Window("NONE",
               Rectangle(vec2(cfg->graphics.res.x - 175,
                              cfg->graphics.res.y - 175),
                         vec2(cfg->graphics.res.x, cfg->graphics.res.y)));
  menu->isVisible(true);
  menu->addMenu("MainMenu",
                { "Master Thesis", "Start Game", "Options", "Exit" },
                { 0, 0 },
                { 15, 50, Menu::VERTICAL, Text::WHITE });


  // Order is important
  mGUIElements = { menu, opts };

  auto handler = [&, opts, menu](const Input::Event& event) {
    if (!menu->isVisible() || menu->isAnimating())
      return;

    menu->defaultInputHandler(event);

    if (event.hasBeenHandled())
      return;

    if (event.keyPressed(GLFW_KEY_ENTER) ||
        event.buttonPressed(GLFW_MOUSE_BUTTON_LEFT)) {
      switch (menu->menu("MainMenu")->getActiveMenu()) {
        case 0:
          event.sendStateChange(States::MasterThesis);
          event.stopPropgation();
          break;
        case 1:
          event.sendStateChange(States::Game);
          event.stopPropgation();
          break;
        case 2:
          opts->isVisible(true);
          menu->isVisible(false);
          event.stopPropgation();
          break;
        case 3:
          event.sendStateChange(States::Quit);
          event.stopPropgation();
          break;
      }
    }
  };

  // add CFG is enabled
  if (cfg->console.enabled) {
    Console* console = new Console(mAsset);
    mLua->add(console);
    mGUIElements.push_back(console);
  }

  menu->setInputHandler(handler);
  mLua->loadFile("lua/states/mainmenu.lua");

  /* mLua->on("reInitialize", [&](){ */
  /*   mLua->add(console); */
  /*   mLua->loadFile("lua/states/mainmenu.lua"); */
  /* }); */
  mLog->debug("Initialized successfully...");
}

MainMenu::~MainMenu() {
  for (auto g : mGUIElements)
    delete g;
}

void MainMenu::update(float deltaTime) {
  mDeltaTime = deltaTime;

  for (auto i = mGUIElements.rbegin(); i != mGUIElements.rend(); ++i) {
    (*i)->update(deltaTime);
  }

  try {
    mLua->engine["update"](deltaTime);
  } catch (const sol::error& e) {
    mLog->error("Failed to update: ", e.what());
  }
}

void MainMenu::draw(float) {
  draw3D();
  drawGUI();
}

void MainMenu::draw3D() {
  glEnable(GL_DEPTH_TEST);
}

void MainMenu::drawGUI() {
  glDisable(GL_DEPTH_TEST);
  for (auto g : mGUIElements)
    g->draw();

  try {
    mLua->engine["draw"]();
  } catch (const sol::error& e) {
    mLog->error("Failed to draw: ", e.what());
  }
}

void MainMenu::input(const Input::Event& event) {
  for (auto i = mGUIElements.rbegin(); i != mGUIElements.rend(); ++i) {
    (*i)->input(event);

    if (event.hasBeenHandled())
      break;
  }

  switch (event.state()) {
    case States::OptionsMenuClose:
      event.sendStateChange(States::NoChange);
      mGUIElements[1]->isVisible(false);
      mGUIElements[0]->isVisible(true);
      break;
  }
}

File: lua/states/mainmenu.lua
-- Engine requires
local vec2            = require 'Math.vec2'
local Window          = require 'GUI.Window'
local TextColor       = require 'GUI.Text.Color'
local MenuOrientation = require 'GUI.Menu.Orientation'
local MenuSettings    = require 'GUI.Menu.Settings'
local Rect            = require 'Shape.Rectangle'
local Util            = require 'Util'
local Keys            = require 'Input.Keys'
local States          = require 'State.Type'

-- File dependencies
local OptionsMenu     = require 'gui/optionsmenu'

local mainMenu = Window.new("NONE", Rect.new(cfg.graphics.res.x - 375,
                                             cfg.graphics.res.y - 175,
                                             cfg.graphics.res.x,
                                             cfg.graphics.res.y))

local menuItems = {"Master Thesis", "Start Game", "Options", "Exit"}
mainMenu:isVisible(true)
local ms = MenuSettings.new(15.0, 50.0, MenuOrientation.Vertical)

mainMenu:addMenu("MainMenu",
                 Util.toVectorStr(menuItems),
                 vec2.new(),
                 ms)


local menu = mainMenu:menu("MainMenu")

local onEvent = function(event)
  if not mainMenu:isVisible() or mainMenu:isAnimating() then
    return
  end

  mainMenu:defaultInputHandler(event)

  if event:hasBeenHandled() then
    return
  end

  local hasConfirm = event:keyPressed(Keys.Enter) or
                     event:buttonPressed(Keys.MouseButton1)

  if not hasConfirm then
    return
  end

  local menuChoose = menu:getActiveMenu()

  if menuChoose == 0 then
    event:sendStateChange(States.MasterThesis)
  elseif menuChoose == 1 then
    event:sendStateChange(States.Game)
  elseif menuChoose == 2 then
  elseif menuChoose == 3 then
    event:sendStateChange(States.Quit)
  end

  event:stopPropgation()
end

mainMenu:setInputHandler(onEvent)

-- addtoDrawable(mainMenu)
